data/transactions/verify/txn_test.go

```diff
@@ -460,14 +663,176 @@ func TestTxnGroupCacheUpdate(t *testing.T) {
+	verifyGroup(t, txnGroups, blkHdr, breakSignatureFunc, restoreSignatureFunc, "rejected by logic")
+
+}
+
+func TestTxnGroupCacheUpdateLogicWithSig(t *testing.T) {
+	partitiontest.PartitionTest(t)
+
+	_, signedTxn, secrets, addresses := generateTestObjects(100, 20, 50)
+	blkHdr := bookkeeping.BlockHeader{
+		Round:       50,
+		GenesisHash: crypto.Hash([]byte{1, 2, 3, 4, 5}),
+		UpgradeState: bookkeeping.UpgradeState{
+			CurrentProtocol: protocol.ConsensusCurrentVersion,
+		},
+		RewardsState: bookkeeping.RewardsState{
+			FeeSink:     feeSink,
+			RewardsPool: poolAddr,
+		},
+	}
+
+	for i := 0; i < len(signedTxn); i++ {
+		op, err := logic.AssembleString(`arg 0
+sha256
+byte base64 5rZMNsevs5sULO+54aN+OvU6lQ503z2X+SSYUABIx7E=
+==`)
+		require.NoError(t, err)
+
+		s := rand.Intn(len(secrets))
+		signedTxn[i].Sig = crypto.Signature{}
+		signedTxn[i].Txn.Sender = addresses[s]
+		signedTxn[i].Lsig.Args = [][]byte{[]byte("=0\x97S\x85H\xe9\x91B\xfd\xdb;1\xf5Z\xaec?\xae\xf2I\x93\x08\x12\x94\xaa~\x06\x08\x849b")}
+		signedTxn[i].Lsig.Logic = op.Program
+		program := logic.Program(op.Program)
+		signedTxn[i].Lsig.Sig = secrets[s].Sign(program)
+
+	}
+
+	txnGroups := make([][]transactions.SignedTxn, len(signedTxn))
+	for i := 0; i < len(txnGroups); i++ {
+		txnGroups[i] = make([]transactions.SignedTxn, 1)
+		txnGroups[i][0] = signedTxn[i]
+	}
+
+	breakSignatureFunc := func(txn *transactions.SignedTxn) {
+		txn.Lsig.Sig[0]++
+	}
+	restoreSignatureFunc := func(txn *transactions.SignedTxn) {
+		txn.Lsig.Sig[0]--
+	}
+	verifyGroup(t, txnGroups, blkHdr, breakSignatureFunc, restoreSignatureFunc, crypto.ErrBatchVerificationFailed.Error())
+}
+
+func TestTxnGroupCacheUpdateLogicWithMultiSig(t *testing.T) {
+	partitiontest.PartitionTest(t)
+
+	secrets, _, pks, multiAddress := generateMultiSigAccounts(t, 30)
+	blkHdr := bookkeeping.BlockHeader{
+		Round:       50,
+		GenesisHash: crypto.Hash([]byte{1, 2, 3, 4, 5}),
+		UpgradeState: bookkeeping.UpgradeState{
+			CurrentProtocol: protocol.ConsensusCurrentVersion,
+		},
+		RewardsState: bookkeeping.RewardsState{
+			FeeSink:     feeSink,
+			RewardsPool: poolAddr,
+		},
+	}
+
+	const numOfTxn = 20
+	signedTxn := make([]transactions.SignedTxn, numOfTxn)
+
+	numMultiSigAcct := len(multiAddress)
+	for i := 0; i < numOfTxn; i++ {
+		s := rand.Intn(numMultiSigAcct)
+		r := rand.Intn(numMultiSigAcct)
+		a := rand.Intn(1000)
+		f := config.Consensus[protocol.ConsensusCurrentVersion].MinTxnFee + uint64(rand.Intn(10))
+
+		signedTxn[i].Txn = transactions.Transaction{
+			Type: protocol.PaymentTx,
+			Header: transactions.Header{
+				Sender:      multiAddress[s],
+				Fee:         basics.MicroAlgos{Raw: f},
+				FirstValid:  basics.Round(1),
+				LastValid:   basics.Round(100),
+				GenesisHash: crypto.Hash([]byte{1, 2, 3, 4, 5}),
+			},
+			PaymentTxnFields: transactions.PaymentTxnFields{
+				Receiver: multiAddress[r],
+				Amount:   basics.MicroAlgos{Raw: uint64(a)},
+			},
+		}
+
+		op, err := logic.AssembleString(`arg 0
```

Comment:
please comment the point of having non-trivial program here

Reply:
I see, it checks it with valid and broken signature.
Maybe add a comment to `TestTxnGroupCacheUpdateLogicWithMultiSig` just to explain the test idea/scenario